src/tools/idlc/CMakeLists.txt

```diff
@@ -9,48 +9,141 @@
 #
 # SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
 #
-configure_file(src/config.h.in config.h)
+include(GenerateExportHeader)
+configure_file(src/idlc/config.h.in config.h)
 
 set(headers
-  src/descriptor.h
-  src/generator.h
-  src/options.h
-  src/plugin.h
-  include/idlc/generator.h
+  include/idlc/plugin.h
   ${CMAKE_CURRENT_BINARY_DIR}/config.h)
 set(sources
-  src/idlc.c
-  src/plugin.c
-  src/options.c
-  src/generator.c
-  src/descriptor.c
-  src/types.c)
-
-if(ENABLE_TYPELIB)
-  list(APPEND headers src/descriptor_type_meta.h)
-  list(APPEND sources src/descriptor_type_meta.c)
-endif()
+  src/idlc/idlc.c
+  src/idlc/plugin.c
+  src/idlc/options.c)
 
 add_executable(idlc ${sources} ${headers})
 
 if(MSVC)
   # ignore warnings C6255 and 6263 about _alloca
   target_compile_options(idlc PRIVATE /wd6255 /wd6263)
 endif()
-target_link_libraries(idlc PRIVATE compat idl idlpp ddsc ${CMAKE_DL_LIBS})
+target_link_libraries(idlc PRIVATE compat idl idlpp ddsc idl_common ${CMAKE_DL_LIBS})
+
 target_include_directories(
   idlc PRIVATE
-    include
+    include/idlc
+    include/generator_common
     ${CMAKE_CURRENT_BINARY_DIR}
     $<BUILD_INTERFACE:$<TARGET_PROPERTY:ddsc,INTERFACE_INCLUDE_DIRECTORIES>>
     $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../../core/ddsi/include>
     $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../../core/cdr/include>)
 
 add_executable(${PROJECT_NAME}::idlc ALIAS idlc)
 
+set(common_hdrs
+  include/generator_common/common.h
+  )
+
+set(common_srcs
+  src/generator_common/common.c
+  )
+
+if(ENABLE_TYPELIB)
+list(APPEND common_hdrs include/generator_common/descriptor_type_meta.h)
+list(APPEND common_hdrs src/generator_common/descriptor_type_meta.c)
+endif()
+
+add_library(
+  idl_common SHARED ${common_hdrs} ${common_srcs})
```

Comment:
So the way I structured things is as follows:
- idlc: the thing that parses the input files and generates the output based on the plugged-in generator
- libcycloneddsidl: parser of idl code to c-structs in a tree
- libidlc_common: common generator functions/structures which are used in multiple language-specific generators (like structs used to communicate, xtypes type/topic information blob generators)
I did it like this since I though it to be weird to have generator libraries linking to eachother, since previously the idlc_common functionality was effectively contained in the code which is now in the c-generator. And requiring/linking the C generator to for instance the C++ generator did not make much sense to me. It also caused issues with multiple definitions of functions with the same name being present in the same library.